packages/firestore/src/platform/config.ts

```diff
@@ -60,15 +58,24 @@ const firestoreNamespace = {
 /**
  * Configures Firestore as part of the Firebase SDK by calling registerService.
  */
-export function configureForFirebase(firebase: FirebaseNamespace): void {
+export function configureForFirebase<T extends Function>(
+  firebase: FirebaseNamespace,
+  publicFirestore: T,
+  internalFirestore: typeof Firestore
+): void {
+  const copiedNamespace = shallowCopy(firestoreNamespace);
+  copiedNamespace.Firestore = publicFirestore;
   (firebase as _FirebaseNamespace).INTERNAL.registerComponent(
     new Component(
       'firestore',
       container => {
         const app = container.getProvider('app').getImmediate()!;
-        return new Firestore(app, container.getProvider('auth-internal'));
+        return new internalFirestore(
```

Comment:
Instead of mandating the shape of the constructor and taking the type as an argument, this contract would be far more flexible if it took a function that took the app and auth providers as arguments and returned a `Firestore` instance. That way the function could do as it pleased, including newing up different internal components to pass to the actual constructor.

As it stands, this makes the means by which the Firestore instance is created very hard to understand.

You might also consider changing the Firestore constructor to take things like the AsyncQueue as parameters itself so that you can just do all the dependency injection uniformly in a single place. Then you end up with two functions:

`newMemoryOnlyFirestore(app, authProvider) => Firestore` that just directly lists out all the items we create including AsyncQueue, the specific PersistenceProvider, etc and a similar one like `newDefaultFirestore` which may duplicate some of that logic but at least it will just be simple constructor dependency wiring without callbacks calling callbacks.